File: 백준/Gold/31965. 회의 장소/회의 장소.py
from bisect import bisect_left, bisect_right
import sys

def get_cost(X, P, i, j, k):
    left_cost = (k - i + 1) * X[k] - (P[k] - P[i - 1])
    right_cost = (P[j] - P[k - 1]) - (j - k + 1) * X[k]
    return left_cost + right_cost

def main():
    input = sys.stdin.readline
    n, q = map(int, input().split())

    X = [0] * (n + 1)
    P = [0] * (n + 1)

    # --- 여기만 수정: X를 여러 줄에서 모아서 n개 채우기 ---
    vals = []
    while len(vals) < n:
        vals.extend(map(int, input().split()))
    for i in range(1, n + 1):
        X[i] = vals[i - 1]
        P[i] = P[i - 1] + X[i]
    # -------------------------------------------------------

    # (필요하다면 보장 차원에서 정렬 전제 확인/정렬)
    # X[1:n+1]이 정렬 전제일 경우 아래 주석 해제 금지. 문제에서 이미 정렬이라고 보장하면 불필요.
    # if not all(X[i] <= X[i+1] for i in range(1, n)):
    #     raise ValueError("X[1..n] must be non-decreasing for bisect to work.")

    for _ in range(q):
        l, r = map(int, input().split())
        lo = bisect_left(X, l, 1, n + 1)
        hi = bisect_right(X, r, 1, n + 1) - 1

        if lo >= hi:
            print(0)
        else:
            max_stress = max(get_cost(X, P, lo, hi, lo),
                             get_cost(X, P, lo, hi, hi))
            min_stress = get_cost(X, P, lo, hi, (lo + hi) // 2)
            print(max_stress - min_stress)

if __name__ == "__main__":
    main()

File: 백준/Gold/31965. 회의 장소/회의 장소.cc
#include <stdio.h>
#include <algorithm>

using namespace std;

int n, q, l, r;
long long X[200010], P[200010];
long long getCost(int i, int j, int k) {
    long long leftCost = (long long)(k - i + 1) * X[k] - (P[k] - P[i - 1]);
    long long rightCost = (P[j] - P[k - 1]) - (long long)(j - k + 1) * X[k];
    
    return leftCost + rightCost;
}

int main() {

    scanf("%d%d", &n, &q);
    

    for (int i = 1 ; i <= n ; ++i) {
        scanf("%lld", &X[i]);
        P[i] = P[i - 1] + X[i];
    }
    

    while (q--) {
        scanf("%d%d", &l, &r);

        int lo = lower_bound(X + 1, X + n + 1, l) - X;
        int hi = upper_bound(X + 1, X + n + 1, r) - X - 1;
        
        if (lo >= hi) {
            puts("0");
        } else {
            long long maxStress = max(getCost(lo, hi, lo), getCost(lo, hi, hi));
            long long minStress = getCost(lo, hi, (lo + hi) / 2);
            
            printf("%lld\n", maxStress - minStress);
        }
    }
}